horovod/spark/common/cache.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import collections
+import threading
+
+
+class TrainingDataCache(object):
+    def __init__(self):
+        self.lock = threading.Lock()
+        self._reset()
+
+    def create_key(self, df, store, validation):
+        return df.__hash__(), store.get_train_data_path(), store.get_val_data_path(), validation
+
+    def set_in_use(self, key, in_use):
+        if in_use:
+            self._keys_in_use[key] += 1
+        else:
+            self._keys_in_use[key] -= 1
+
+    def create_data_paths(self, key, store):
+        _, _, _, validation = key
+        idx = 0
```

Comment:
We do have support for this using the `fit_on_parquet` method of the estimator.  It does require that the user setup their store to explicitly reference the correct saved DataFrame path.

Databricks is working on an API that will simplify the DataFrame -> Petastorm Parquet conversion so that it should be easier to perform in two steps.  Though as you mentioned, users can choose to use `prepare_data` for now.